core/matchers/src/main/kotlin/au/com/dius/pact/core/matchers/JsonBodyMatcher.kt

```diff
@@ -132,6 +175,22 @@ object JsonBodyMatcher : BodyMatcher, KLogging() {
         }
         result
       }
+    } else {
+      // match unordered list
+      logger.debug { "compareLists: IgnoreOrderMatcher defined for path $path" }
+      val result = Matchers.domatch(matchers, "body", path, expectedValues, actualValues, BodyMismatchFactory)
+          .toMutableList()
+      if (expectedList.isNotEmpty()) {
+        // No need to pad 'expected' as we already visit all 'actual' values
+        result.addAll(compareListContentUnordered(expectedList, actualList, path, allowUnexpectedKeys, matchers))
+        if (Matchers.ignoreOrderMatcherDefined("body", path, matchers) &&
+            (expectedValues.size() != actualValues.size())) {
+          result.add(BodyMismatch(a, b,
+            "Expected a List with ${expectedValues.size()} elements but received ${actualValues.size()} elements",
+            path.joinToString("."), generateJsonDiff(expectedValues, actualValues)))
+        }
+      }
+      result
```

Comment:
I can see your reasoning for reordering this.  My only hesitation is that it follows the same order as it did previously and just tacks ignore-order onto the end.  I could make ignore-order first like so...

```
return if (Matchers.isEqualsIgnoreOrderMatcherDefined("body", matchers)) {
    // match unordered list
} else if (Matchers.matcherDefined("body", path, matchers)) {
  // match ordered list - at least one body matcher defined
} else {
  // match ordered list - no body matcher, default to equals matcher
}
```

Reply:
Actually that order failed on the test with one ordered and one unordered...

```
  def 'matching json bodies - return a mismatches - with ignore-order set on another list'() {
    given:
    matchers.addCategory('body')
      .addRule('$[0].array1', EqualsIgnoreOrderMatcher.INSTANCE)

    expect:
    matcher.matchBody(expectedBody, actualBody, true, matchers).find {
      it instanceof BodyMismatch && it.mismatch.contains('Expected 1 but received 2')
    }

    where:
    actualBody = OptionalBody.body('''{
     "array1": [{"foo": "a"},{"foo": "b"}],
     "array2": [2, 3, 1, 4]
     }'''.bytes)
    expectedBody = OptionalBody.body('''{
     "array1": [{"foo": "b"},{"foo": "a"}],
     "array2": [1, 2, 3, 4]
     }'''.bytes)
  }
```

Reply:
Does `oneOfIgnoreOrderMatchersDefined` not take a path, or did you just omit it from your code snippet?  It would need to the path, especially in a situation like this test.

Reply:
Well it did, but it wasn't being used, so I removed it.  When I tried this, I thought about adding it back and testing using matchesPath like wildcardMatcherDefined does.  

However, I don't think that would work because it filters on path.size which is the same for both array1 and array2 in this case.

```
  @JvmStatic
  fun wildcardMatcherDefined(path: List<String>, category: String, matchers: MatchingRules?) =
    if (matchers != null) {
      val resolvedMatchers = matchers.rulesForCategory(category).filter(Predicate {
        matchesPath(it, path) == path.size
      })
      resolvedMatchers.matchingRules.keys.any { entry -> entry.endsWith(".*") }
    } else false
```

Reply:
I'll give it shot tho.

Reply:
So, I made the change to filter on path length and that broke some tests in JsonBodyMatcherSpec.  

Originally I was looking at 'isEqualsIgnoreOrderMatcherDefined' and 'isOtherThanEqualsIgnoreOrderMatcherDefined' as being not path based.  Now I'm wondering if they should be based on the path.  However, the length alone is not specific enough because that will generate a lot of false positives.

I'm going to have to think about this a little more.

Reply:
I was able to fix this.  Turned out well, because it identified another couple of problems that would have cropped up later.